From 8b76818914fe0288eca39e82a90143117fb23c7c Mon Sep 17 00:00:00 2001
From: Wyon Bi <bivvy.bi@rock-chips.com>
Date: Thu, 7 Sep 2017 14:13:19 +0800
Subject: [PATCH] drm/panel: simple: Add support for LG LM238WR2-SPA1

This adds support for the LG Corporation LM238WR2-SPA1 23.8"
eDP(HBR2, 5.4Gbps) UHD TFT LCD panel, which can be supported
by the simple panel driver.

Change-Id: I586684ba893be54fbf664a8fa4dad0fe5eb999a0
Signed-off-by: Wyon Bi <bivvy.bi@rock-chips.com>
---
 .../display/panel/lg,lm238wr2-spa1.txt        |  9 +++++
 drivers/gpu/drm/panel/panel-simple.c          | 33 +++++++++++++++++++
 2 files changed, 42 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/display/panel/lg,lm238wr2-spa1.txt

diff --git a/Documentation/devicetree/bindings/display/panel/lg,lm238wr2-spa1.txt b/Documentation/devicetree/bindings/display/panel/lg,lm238wr2-spa1.txt
new file mode 100644
index 000000000000..00e5765a3735
--- /dev/null
+++ b/Documentation/devicetree/bindings/display/panel/lg,lm238wr2-spa1.txt
@@ -0,0 +1,9 @@
+LG Corporation 23.8" UHD TFT LCD panel
+
+It has been designed to apply eDP(HBR2, 5.4Gbps) interface.
+
+Required properties:
+- compatible: should be "lg,lm238wr2-spa1"
+
+This binding is compatible with the simple-panel binding, which is specified
+in simple-panel.txt in this directory.
diff --git a/drivers/gpu/drm/panel/panel-simple.c b/drivers/gpu/drm/panel/panel-simple.c
index e509c16abe3f..2e8b210fe74e 100644
--- a/drivers/gpu/drm/panel/panel-simple.c
+++ b/drivers/gpu/drm/panel/panel-simple.c
@@ -1360,6 +1360,36 @@ static const struct panel_desc sharp_lcd_f402 = {
 	.bus_format = MEDIA_BUS_FMT_RGB666_1X18,
 };
 
+static const struct drm_display_mode lg_lm238wr2_spa1_mode = {
+	.clock = 533250,
+	.hdisplay = 3840,
+	.hsync_start = 3840 + 48,
+	.hsync_end = 3840 + 48 + 32,
+	.htotal = 3840 + 48 + 32 + 80,
+	.vdisplay = 2160,
+	.vsync_start = 2160 + 3,
+	.vsync_end = 2160 + 3 + 5,
+	.vtotal = 2160 + 3 + 5 + 54,
+	.vrefresh = 60,
+	.flags = DRM_MODE_FLAG_NVSYNC | DRM_MODE_FLAG_NHSYNC,
+};
+
+static const struct panel_desc lg_lm238wr2_spa1 = {
+	.modes = &lg_lm238wr2_spa1_mode,
+	.num_modes = 1,
+	.bpc = 8,
+	.size = {
+		.width = 527,
+		.height = 297,
+	},
+	.delay = {
+		.prepare = 20,
+		.enable = 20,
+		.unprepare = 20,
+		.disable = 20,
+	},
+};
+
 static const struct drm_display_mode lg_lm270wr3_ssa1_mode = {
 	.clock = 533250,
 	.hdisplay = 3840,
@@ -1719,6 +1749,9 @@ static const struct of_device_id platform_of_match[] = {
 	}, {
 		.compatible = "lg,lb070wv8",
 		.data = &lg_lb070wv8,
+	}, {
+		.compatible = "lg,lm238wr2-spa1",
+		.data = &lg_lm238wr2_spa1,
 	}, {
 		.compatible = "lg,lm270wr3-ssa1",
 		.data = &lg_lm270wr3_ssa1,
-- 
2.35.3

